Allow for queries on _tier to skip shards during coordinator rewrite

andreidan opened the pull request:
The `_tier` metadata field was not used on the 
coordinator when rewriting queries in order to
exclude shards that don't match. This lead to queries
in the following form to continue to report failures
even though the only unavailable shards were in the 
tier that was excluded from search (frozen tier in this
example):
```
POST testing/_search
{
  "query": {
    "bool": {
      "must_not": [
        {
          "term": {
            "_tier": "data_frozen"
          }
        }
      ]
    }
  }
}
```

This PR addresses this by having the queries that can
execute on `_tier` (term, match, prefix, wildcard) execute
a coordinator rewrite to exclude the indices that don't match
the `_tier` query **before** attempting to reach to the shards 
(shards, that might not be available and raise errors). 

Fixes #114910

andreidan committed:
Add tier to CoordinatorRewriteContext and test to query missing partially mounted index

andreidan committed:
The _tier field is exposed in server

andreidan committed:
Only match _tier on the coordinator if _tier specified in term query

andreidan committed:
Fix tests

andreidan committed:
rename getter

andreidan committed:
minor refactoring and tests

andreidan committed:
minor refactoring and tests

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into can-match-doesnt-fail-on-frozen

andreidan committed:
Add bool query and match_all when _tier matches

andreidan committed:
spotless

andreidan committed:
trim things

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into can-match-doesnt-fail-on-frozen

andreidan committed:
More queries and tests

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into can-match-doesnt-fail-on-frozen

andreidan committed:
spotless

andreidan committed:
This is not needed as tier mapper tests are now in server

andreidan commented on x-pack/plugin/core/src/main/java/module-info.java:
```diff
@@ -34,7 +34,6 @@
     exports org.elasticsearch.protocol.xpack;
     exports org.elasticsearch.snapshots.sourceonly;
     exports org.elasticsearch.xpack.cluster.action;
-    exports org.elasticsearch.xpack.cluster.routing.allocation.mapper;
```
This is now in `server`

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java:
```diff
@@ -303,11 +301,6 @@ private static boolean alreadyContainsXPackCustomMetadata(ClusterState clusterSt
             || metadata.custom(TransformMetadata.TYPE) != null;
     }
 
-    @Override
```
Moved to `server`

elasticsearchmachine commented:
Pinging @elastic/es-search-foundations (Team:Search Foundations)

andreidan committed:
Update docs/changelog/114990.yaml

elasticsearchmachine commented:
Hi @andreidan, I've created a changelog YAML for you.

andreidan commented:
Opened for an initial review. Unit tests on the various queries we're changing here are left to add (new code is covered by an integration test for now).

andreidan commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContextProvider.java:
```diff
@@ -74,7 +74,8 @@ public CoordinatorRewriteContext getCoordinatorRewriteContext(Index index) {
             nowInMillis,
-            new DateFieldRangeInfo(timestampFieldType, timestampRange, dateFieldRangeInfo.eventIngestedFieldType(), eventIngestedRange)
+            new DateFieldRangeInfo(timestampFieldType, timestampRange, dateFieldRangeInfo.eventIngestedFieldType(), eventIngestedRange),
+            indexMetadata.getTierPreference().isEmpty() == false ? indexMetadata.getTierPreference().getFirst() : ""
```
We are potentially artificially limiting the coordinator tier rewrite to [only indices that have the time range configured](https://github.com/elastic/elasticsearch/pull/114990/files#diff-0ad1800461d83a801b07a1207f4395615a0e9b6b5aad9ded33bda88e0f784d47R56) (partially mounted indices). We don't have to have this limitation for the `_tier` coordination rewrite however, the current implementation is sufficient for skipping frozen (where all indices do indeed have the `@timestamp` and `event.ingested` ranges). 

In the interest of keeping this fix targeted and contained I believe it's a good compromise for now but curious what you think.

javanna replied:
That's quite subtle, is that because above we return null if the range field info isn't found in the cluster state? 

```
DateFieldRangeInfo dateFieldRangeInfo = mappingSupplier.apply(index);
if (dateFieldRangeInfo == null) {
    return null;
}
```

I think this is ok, but perhaps let's clarify in the code that it's by design and not by mistake? Perhaps we will want to extend this in the future to all indices? I mean there is no technical reason that this tier filtering cannot be extended to all indices, provided that the tier field cannot ever change (which may happen for some constant field types - e.g. they may be unset and get set for the first time via dynamic mappings)

javanna replied:
now that this is merged, do we want to do the follow-up to allow filtering on tier to work on all indices?

andreidan replied:
Yes, PR is up https://github.com/elastic/elasticsearch/pull/115797

andreidan committed:
unit tests

piergm reviewed (commented):
Changes make sense to me and form a first pass everything looks good. The only thing I wonder is if there is a way to avoid all the *QueryBuilder#doCoordinatorRewrite overrides, probably there is not, but that would make the change much simpler.
I am saying this because for @timestamp was a smaller change, since we can "only" use RangeQueryBuilder, while for _tier we could use a many more types of queries.

javanna commented on server/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java:
```diff
@@ -968,6 +969,22 @@ protected Query doToQuery(SearchExecutionContext context) throws IOException {
     }
 
+    @Override
+    protected QueryBuilder doCoordinatorRewrite(CoordinatorRewriteContext coordinatorRewriteContext) {
```
I see that query_string does not override any of the rewrite methods yet. I was wondering how querying multiple fields is combined in a single lucene query depending on the operators, and I think that we can't state for sure that if one clause is on the tier field we can rewrite to match_none or match_all.

In general, the approach I would take is the following: if queries have doIndexMetadataRewrite with logic on constant field type, we can do the same in coordinating rewrite for the data tier field. Queries that don't do that based on mappings, shouldn't do it on the coord node either.

andreidan replied:
Ah great shout! Sounds good, I'll reduce the scope here. Thanks !

javanna commented on server/src/main/java/org/elasticsearch/index/query/SimpleQueryStringBuilder.java:
```diff
@@ -609,6 +610,21 @@ public static SimpleQueryStringBuilder fromXContent(XContentParser parser) throw
+            return this;
+        }
+    }
+
```
similar to query_string, I am not too sure that we should do this for simple query string

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DataTierFieldMapper.java:
```diff
@@ -1,11 +1,13 @@
  */
 
-package org.elasticsearch.xpack.cluster.routing.allocation.mapper;
+package org.elasticsearch.index.mapper;
```
I wonder if moving the tier field to core is something that should be done as a prerequisite follow-up. I had almost missed it among other changes here. Was it a requirement for your change?

andreidan replied:
This was needed due to the change in `CoordinatorRewriteContext#getFiledType` https://github.com/elastic/elasticsearch/pull/114990/files#diff-45c9ed5af87237215a5a11f3d2549aa678272c2cbc898b866b43516ec992fab5R86

javanna replied:
makes sense, we could have probably hacked as well by doing string comparison and creating a custom constant  field type instance? Playing devil's advocate a bit here. Or perhaps do the move as a pre-requisite in a separate PR?

javanna reviewed (commented):
I left some comments, this looks pretty good though

andreidan commented on server/src/main/java/org/elasticsearch/index/query/PrefixQueryBuilder.java:
```diff
@@ -234,6 +235,25 @@ protected Query doToQuery(SearchExecutionContext context) throws IOException {
 
+    @Override
+    protected QueryBuilder doCoordinatorRewrite(CoordinatorRewriteContext coordinatorRewriteContext) {
+        if (fieldName.equals(DataTierFieldMapper.NAME) == false) {
```
This is the field the query executes on
e.g. 
```
"query": {
    "bool": {
      "must_not": [
        {
          "term": {
  ****          "_tier": "data_frozen" **** this field
          }
        }
      ]
    }
```

This is to ensure that we **don't** do tier filtering if `_tier` is not specified in the query
e.g. this query will _not_ do any coordinator rewrite
"query": {
    "bool": {
      "must_not": [
        {
          "term": {
              "other_field": "some_value"
          }
        }
      ]
    }
```

Is this an incorrect invariant?

andreidan replied:
Ah, I noticed the simplification here :) 
`coordinatorRewriteContext.getFieldType` will return `null` if the query is not targeting the `_tier` field 👍

andreidan replied:
Ah, I noticed the simplification here :)
coordinatorRewriteContext.getFieldType will return null if the query is not targeting the _tier field 👍

javanna commented on server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java:
```diff
@@ -311,6 +314,9 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
         }
         final QueryRewriteContext context = queryRewriteContext.convertToIndexMetadataContext();
         if (context != null) {
+            // Even though we're still on the coordinator, this executes at the shard level.
```
it is the search_coordinator thread pool in the data node :)

andreidan replied:
Ah of course!! Thanks ! I know nothing. :)

andreidan commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContext.java:
```diff
@@ -99,4 +108,8 @@ public IndexLongFieldRange getFieldRange(String fieldName) {
+
+    public String tier() {
+        return tier;
+    }
```
I believe we need the `tier` as the `CoordinatorRewriteContext` doesn't know of `indexSettings` otherwise https://github.com/elastic/elasticsearch/pull/114990/files/67f82e9aa742e223de4cd6a76d496ac66987130e#diff-45c9ed5af87237215a5a11f3d2549aa678272c2cbc898b866b43516ec992fab5R60
The index settings would be needed for us to parse the `_tier_preference` setting.

I'm not sure if it should ? Or if we should keep just passing the `tier` ?

javanna replied:
I was under the impression that that is the same value that the constant field type that the context exposes returns, or is that something else?

andreidan replied:
The field type does not store any state but every time it evaluates it reaches out to the index settings for the `_tier_preference` value https://github.com/elastic/elasticsearch/pull/114990/files/67f82e9aa742e223de4cd6a76d496ac66987130e#diff-e776fbfb85ffa7c76ad9fbbd1263953047589070bf15add46298deaf929778dbR95

The `QueryWriteContext` stores the index settings (but the `CoordinatorRewriteContext` doesn't)

javanna replied:
sounds good, I had misunderstood.

andreidan committed:
Remove false doc

andreidan committed:
Reuse the                               Reuse the doIndexMetadatRewrite logic to query constant fields

javanna commented:
>  The only thing I wonder is if there is a way to avoid all the *QueryBuilder#doCoordinatorRewrite overrides, probably there is not, but that would make the change much simpler.

That looks ok to me, it's not too many queries, and that makes it super clear which queries support which rewrite mechanism.

andreidan committed:
Only rewrite missing fields to match_none on data nodes

Mappings are not materialised on the coordinator node so
we can't distinguish a valid mapping from a missing field.
Currently, the only available mappings on the coordinator
are @timestamp, event.ingested and _tier

andreidan commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

andreidan committed:
Merge branch 'main' into can-match-doesnt-fail-on-frozen

andreidan committed:
javadoc

andreidan committed:
New api is happy

andreidan committed:
Document _tier rewrite limitation

andreidan committed:
Revert moving the datatier field to server

andreidan committed:
CoordinatorRewriteContext defined tier field type itself

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into can-match-doesnt-fail-on-frozen

andreidan commented:
@piergm 

> The only thing I wonder is if there is a way to avoid all the *QueryBuilder#doCoordinatorRewrite overrides

Apologies for the delay. I believe this is a nice feature at the moment. It's currently quite straightforward to flush out the queries that perform a coordinator rewrite

andreidan committed:
spotless

piergm reviewed (approved):
LGTM!
Kudos on all the comments explaining what you are doing in tricky parts of the code, our future selfs will love that!
And extra kudos on re-using part of the metadata rewrite logic, this makes things really clean IMO! 😄 👏

javanna commented on server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java:
```diff
@@ -380,6 +384,43 @@ protected QueryBuilder doIndexMetadataRewrite(QueryRewriteContext context) throw
+        // If we're using a keyword analyzer then we can rewrite this to a TermQueryBuilder
+        // and possibly shortcut
+        NamedAnalyzer configuredAnalyzer = configuredAnalyzer(coordinatorRewriteContext);
+        if (configuredAnalyzer != null && configuredAnalyzer.analyzer() instanceof KeywordAnalyzer) {
```
I don't think we can reliably do this for match query in the coord node. The analyzer bits come from mappings (see configuredAnalyzer) that are not necessarily available on the coord node. There is one case where the query may specify an analyzer, but still that analyzer needs to be looked up via mapper service which is going to be null on the coord node if no shards for the queried index are allocated on it. I think that there is no way around it.

andreidan replied:
Ah great catch. Removing now

javanna reviewed (approved):
This looks great! I left a rather important comment on the logic for match query, but I am good with this going in without additional reviews once the last comments are addressed

andreidan committed:
Make static method an instance method

andreidan commented on server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:
```diff
@@ -407,4 +410,27 @@ public ResolvedIndices getResolvedIndices() {
+     * present, then return null.
+     */
+    @Nullable
+    public static String getTierPreference(QueryRewriteContext context) {
```
Sounds good. Thank you!

andreidan commented on server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:
```diff
@@ -407,4 +410,27 @@ public ResolvedIndices getResolvedIndices() {
+            return tier.isEmpty() == false ? tier : null;
+        }
+        Settings settings = context.getIndexSettings().getSettings();
+        String value = DataTier.TIER_PREFERENCE_SETTING.get(settings);
```
We've had performance issues in the past with parsing the tier preference from the settings. We now have a version of the `IndexMetadata` that does the parsing and caches the result for the lifetime of the `IndexMetadata`. 
Since the `CoordinatorRewriteContextProvider` has access to the `IndexMetadata` I preferred to take this route rather than re-parsing the setting.

andreidan commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContextProvider.java:
```diff
@@ -52,6 +52,12 @@ public CoordinatorRewriteContext getCoordinatorRewriteContext(Index index) {
+        // rewrite only based on the _tier field) and we might decide to remove this artificial
+        // limitation to enable coordinator rewrites based on _tier for hot and warm indices
+        // (currently the _tier coordinator rewrite is only available for mounted and partially mounted
+        // indices)
```
I believe read-only indices is a superset of where we actually have this optimization available - i.e. only for fully and partially mounted indices ?
I believe the comment is accurate if my assumption is correct (i.e. hot/warm indices don't have this optimization, cold/frozen ones - fully/partially mounted ones - do) but happy to clarify further if you have a better suggestion?

andreidan commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContext.java:
```diff
@@ -30,20 +38,66 @@
+            if (tierPreference == null) {
+                return new MatchNoDocsQuery();
+            }
+            return new MatchAllDocsQuery();
```
Ah good shout. Done

andreidan committed:
UnsupportedOperationException where needed

andreidan committed:
Remove coordinator rewrite support from match query

andreidan committed:
spotless

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into can-match-doesnt-fail-on-frozen

andreidan committed:
Drop empty line

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.16](https://github.com/elastic/elasticsearch/pull/115513)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/115513">](https://github.com/elastic/elasticsearch/pull/115513) |
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/115514)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/115514">](https://github.com/elastic/elasticsearch/pull/115514) |
